File: DAG_C_Interface/examples/3d7p/3d7p_dag.f
!****************************************************
!  task partition only on two dimensions <Z,Y-split,X-split>
!                       i.e. in loop split<k,j-split,i-split> 
!  usage: make see fdag in makefile, run as ./3d7p 
!***************************************************
#define alpha 0.0876
#define beta 0.0765

     module stencil_dag_related
     use  ISO_C_BINDING
     use dag_struct
     real(c_double),allocatable ::p(:,:,:,:)
     real(c_double),allocatable ::t(:,:,:,:)
     type ARG_TYPE    
       integer(c_int) starti,endi,startj,endj
       integer(c_int) sizeK
       integer(c_int) dest,src
     end type
     end module stencil_dag_related

     program main
     use  ISO_C_BINDING
     use  stencil_dag_related
     implicit none
     ! fortran array, default lower bound is 1
#include "aceMesh_runtime_f.h"
    INTERFACE
     subroutine neighbor(neighbor_addrs,args) 
       use  stencil_dag_related
      implicit none
      type(ARG_TYPE),target :: args
      type(ptr_array),intent(inout)::neighbor_addrs
     END
     subroutine core_3d7p(args)
       use  stencil_dag_related
      implicit none
      type(ARG_TYPE),target :: args
     END
    integer function check(iter)
     use  stencil_dag_related
     implicit none
        integer(c_int) iter
    end
    END INTERFACE
     external timer_read
     real (c_double) timer_read,tt
     integer(c_int) num_thread,ret
     character(3)  thread
     integer(c_int) itt,i,j,endj,endi,num 
     integer(c_int) arg_size,dest,src
     type(ARG_TYPE) :: each
     integer(c_int),pointer :: pp =>null()
     allocate (p(SIZEZ,SIZEY,SIZEX,2))
    allocate (t(SIZEZ,SIZEY,SIZEX,2))
     !num_thread = 1
     call my_init()
     !****************************************************************
     ! we suggest use ACEMESH_NUM_THREADS as specified by dag library
     ! we do not encourage to specify num_threads in the source program
     ! we init using argument=0
     !****************************************************************
     !      delete the following lines!
     !call get_environment_variable("NUM_THREADS",thread)
     !read (thread,'(I3)') num_thread
     !***************************************************************
     call timer_clear(1)
     call acemesh_runtime_init(0)
     
     call timer_start(1)
     !num=1
     do itt=1,ITER
        !the first iteration, dest=2
        !dest,src in[1,2]
        dest=mod(itt, 2)+1
        src=mod(itt+1,2)+1
        call acemesh_begin_split_task('3d7p')
        do j=2,SIZEY-1,BLKY
          if(j+BLKY-1 > SIZEY-1)then
            endj=SIZEY-1
          else
            endj=j+BLKY-1
          endif
          do i=2,SIZEX-1,BLKX
            if(i+BLKX-1>SIZEX-1)then
              endi=SIZEX-1
            else
              endi=i+BLKX-1
            endif
            each%starti=i
            each%endi=endi
            each%startj=j
            each%endj=endj
            each%sizeK=SIZEZ
            each%dest=dest
            each%src=src
            arg_size = sizeof(each)
            call acemesh_push_wlist1(1,c_loc(p(1,i,j,dest)),NORMAL)
            call acemesh_push_rlist1(1,c_loc(p(1,i,j,src)),NORMAL)
            call acemesh_task_generator_with_neighbors(c_funloc(core_3d7p),    &
                  c_loc(each),arg_size,c_loc(pp),c_funloc(neighbor),c_loc(each))
            call acemesh_task_set_type(STENCIL_TASK)
          enddo
        enddo
        call acemesh_end_split_task()
     enddo ! time iteration
     call acemesh_spawn_and_wait(99)
     call timer_stop(1)
     
     call acemesh_runtime_shutdown()
     tt=timer_read(1)
     ret = check(ITER)
     if(ret == -1)then
        print *,'error'
     else
        print *,'dag time including graph build:',tt
        print *,'check ok!'
     endif
    deallocate(p)
    deallocate(t)
!    deallocate(c_array)
     END
     !contain
      !*****************************************
      !  neighbor addrs
      !*****************************************
      subroutine neighbor(neighbor_addrs,args)
    use  stencil_dag_related
    use ISO_C_BINDING 
       implicit none
#include "aceMesh_runtime_f.h"
      integer(c_int) starti,startj,sz,i
      type(ARG_TYPE),target :: args
      type(ptr_array) :: neighbor_addrs
      integer(c_int),target :: num_neigh
      integer(c_int) dest,src,num
      type(c_ptr),pointer::array(:)
      sz=MAX_NEIGHBORS
      call c_f_pointer(neighbor_addrs%arr,array,[sz])
      starti = args%starti
      startj = args%startj
      src=args%src
      num=0;
      if (starti>=BLKX+2)then
         num=num+1
         array(num) = c_loc(p(1,starti-BLKX,startj,src))
         !write(*,"(1x, Z12)") loc(u(1,startj,starti-BLKX,src))
      endif
      if (starti<SIZEX-BLKX)then
         num=num+1
         array(num) = c_loc(p(1,starti+BLKX,startj,src))
         !write(*,"(1x, Z12)") loc(u(1,startj,starti+BLKX,src))
      endif
      if (startj>=BLKY+2)then
         num=num+1
         array(num) = c_loc(p(1,starti,startj-BLKY,src)) 
         !write(*,"(1x, Z12)") loc(u(1,startj-BLKY,starti,src))
      endif
      if (startj<SIZEY-BLKY)then
         num=num+1
         array(num) = c_loc(p(1,starti,startj+BLKY,src)) 
         !write(*,"(1x, Z12)") loc(u(1,startj+BLKY,starti,src))
      endif
      neighbor_addrs%len=num
      return
     end subroutine neighbor

     !**********************************
     !  task's computation code
     !**********************************
      subroutine core_3d7p(args)
    use  stencil_dag_related
    use ISO_C_BINDING
      implicit none
      type(ARG_TYPE),target :: args
     integer(c_int) starti,endi,startj,endj
     integer(c_int) i,j,k,sizek,dest,src
     starti = args%starti
     endi = args%endi
     startj = args%startj
     endj = args%endj
     dest = args%dest
     src = args%src
     sizek = args%sizeK
     !NOTE: you can not print to stdout here,  
     !forrtl: severe (40): recursive I/O operation, unit -1, file unknown
     do i=starti,endi
     do j=startj,endj
     do k=2, sizek - 1
        p(k,j,i,dest) = alpha * p(k,j,i,src) +beta * (p(k,j,i+1,src) + &
                p(k,j+1,i,src)+ p(k+1,j,i,src)+ p(k,j,i-1,src) &
                + p(k,j-1,i,src)+ p(k-1,j,i,src))
     enddo
     enddo
     enddo
     return
     end subroutine core_3d7p
    
!**********************************************
!     data initialization function 
!**********************************************
     subroutine my_init()
    use  stencil_dag_related
     implicit none
     integer(c_int) i,j,k
     do k=1,SIZEX
     do j=1,SIZEY
     do i=1,SIZEZ
         p(i,j,k,1)= (k*2.5 + j*3.3 + i*0.5)/3.3
         t(i,j,k,1)=p(i,j,k,1)
     enddo
     enddo
     enddo
     !array2's boundary data
     do k=1,SIZEX,SIZEX-1
     do j=1,SIZEY,SIZEY-1
     do i=1,SIZEZ,SIZEZ-1
         p(i,j,k,2)= (k*2.5 + j*3.3 + i*0.5)/3.3
         t(i,j,k,2)=p(i,j,k,2)
     enddo
     enddo
     enddo

     return
     end subroutine my_init

!**********************************************
!     check function 
!**********************************************
     integer function check(iter)
     use  stencil_dag_related
     implicit none
     integer(c_int) iter,i,j,k,itt,dest,src
     integer(c_int) fab
         
     dest=1  !this is for special case of zero iteration stencil
     do  itt = 1, iter 
       dest=mod(itt,2)+1
       src =mod(itt+1,2)+1

       do k=2, SIZEX-1
       do j=2, SIZEY-1
       do i=2, SIZEZ-1
          t(i,j,k,dest) = alpha * t(i,j,k,src) +beta * (t(i,j,k+1,src) +  &
                t(i,j+1,k,src)+ t(i+1,j,k,src)+ t(i,j,k-1,src)+ t(i,j-1,k,src)+t(i-1,j,k,src))
       enddo
       enddo
       enddo
     enddo    
     !check...
     do k=2,SIZEX-1
     do j=2,SIZEY-1
     do i=2,SIZEZ-1
        fab=abs(t(i,j,k,dest)-p(i,j,k,dest))
        if(fab >= 1e-16)then
            print *,'i=',i,'j=',j,'k=',k
            print *,t(i,j,k,dest),'vs',p(i,j,k,dest)
            print *,'different!'
            check = -1
            return
        endif
     enddo
     enddo
     enddo
     check = 0
     return
     end function check

     !end program main

File: DAG_C_Interface/examples/3d7p_chuanxing/reuse_test/3d7p_dag.c
//you should define SIZEX/SIZEY/SIZEZ/BLKX/BLKY/ITER
//actually 2d-partitioning in this program 

//#include <iostream>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <assert.h>

#include <sys/time.h>
#include "aceMesh_runtime_c.h"

double A[2][SIZEX][SIZEY][SIZEZ]; //__attribute__ ((aligned(64)));
//for test
double test[2][SIZEX][SIZEY][SIZEZ];
int BLKX, BLKY;

//arg
#define alpha_d 0.0876
#define beta_d  0.0765
//kernel 
#define Do3d7pkernel(dest,src,i,j,k) {\
   A[dest][i][j][k] = alpha_d * (A[(src)][(i)][(j)][(k)]) + \
                        beta_d * (A[src][i-1][j][k] + A[src][i+1][j][k] +\
                        A[src][i][j-1][k] + A[src][i][j+1][k] +\
                        A[src][i][j][k-1] + A[src][i][j][k+1]);}
//
#define TestDo3d7pkernel(itr,i,j,k) {\
   test[itr%2][i][j][k] = alpha_d * test[(itr+1)%2][i][j][k] + \
                        beta_d * (test[(itr+1)%2][i-1][j][k] + test[(itr+1)%2][i+1][j][k] +\
                        test[(itr+1)%2][i][j-1][k] + test[(itr+1)%2][i][j+1][k] +\
                        test[(itr+1)%2][i][j][k-1] + test[(itr+1)%2][i][j][k+1] ); } 

typedef struct{
  int starti, endi;
  int startj, endj;
  int k_N;
  int iter;
  void* src_addr;
}targs;
    
void my_neighbors(struct ptr_array *neighbor_addrs, targs* args)
{
    int starti=args->starti;
    int startj=args->startj;
    void* src_addr=args->src_addr;
  
    neighbor_addrs->len=0;
    if(starti - BLKX>=0)
        neighbor_addrs->arr[(neighbor_addrs->len)++]=(void*)(src_addr - BLKX * SIZEY * SIZEZ * 8);
    if(starti + BLKX < SIZEX)
        neighbor_addrs->arr[(neighbor_addrs->len)++]=(void*)(src_addr + BLKX * SIZEY * SIZEZ * 8);
    if(startj - BLKY >= 0)
        neighbor_addrs->arr[(neighbor_addrs->len)++]=(void*)(src_addr - BLKY * SIZEZ * 8);
    if(startj + BLKY < SIZEY)
        neighbor_addrs->arr[(neighbor_addrs->len)++]=(void*)(src_addr + BLKY * SIZEZ* 8);
}
void stencil_core(targs* args)
{
    int starti=args->starti;
    int endi=args->endi;
    int startj=args->startj;
    int endj=args->endj;
    int iter=args->iter;
    int i,j,k;

    int dest = iter  % 2 ;
    int src = 1 - dest;
    for(i = starti; i < endi; i++)
      for(j = startj; j < endj; j++)
        for(k = 1; k < args->k_N - 1; k++) 
          Do3d7pkernel(dest,src,i,j,k);
}


int Do3d7p(const int bkn, const int iter, const int blocsizex, const int blocsizey)
{
    targs each;
	int itr,i,j;
    const int ntaski = SIZEX / blocsizex;
    const int ntaskj = SIZEY / blocsizey;

    for(itr = 0; itr < iter; ++itr)
    {
        acemesh_begin_split_task("3d7p iter");
        for(i = 1; i < SIZEX -1; i += blocsizex)
        {
            int endi = (i + blocsizex > SIZEX -1 ) ? SIZEX - 1 : i + blocsizex;
            int indexi = (i-1) / blocsizex;
            for(j = 1; j < SIZEY -1; j += blocsizey)
            {
                int indexj = (j-1) / blocsizey;
                int endj = (j + blocsizey > SIZEY -1 ) ? SIZEY - 1 : j + blocsizey;
                each.starti=i;
                each.endi=endi;
                each.startj=j;
                each.endj=endj;
                each.k_N=bkn;
                each.iter=itr;
                each.src_addr=&A[(itr+1)%2][i][j][0];
                acemesh_push_wlist(1, &A[itr%2][i][j][0],NORMAL);
                acemesh_push_rlist(1, &A[(itr+1)%2][i][j][0], NORMAL);
                acemesh_task_generator_with_neighbors((TASK_FUNCPTR)stencil_core,
                      (void*)(&each), sizeof(targs), 
                      NULL, (NEIGHBOR_FUNCPTR)my_neighbors, &each);
                acemesh_task_set_type(STENCIL_TASK);
            }
        }
        acemesh_end_split_task();
    }
    acemesh_spawn_and_wait(1);
    //acemesh_wait_for_all_task();
    return 0;
}



void Myinit()
{
    int i,j,k;
    for(int i = 0; i < SIZEX; i++)
    for(int j = 0; j < SIZEY; j++)
    for(int k = 0; k < SIZEZ; k++)
    {
      A[1][i][j][k] = (double)(i*2.5 + j*3.3 + k*0.5 ) /3.0;
      test[1][i][j][k]=A[1][i][j][k];
    }
	//init boundary data of array[0]
    for(int i = 0; i < SIZEX; i+=SIZEX-1)
    for(int j = 0; j < SIZEY; j+=SIZEY-1)
    for(int k = 0; k < SIZEZ; k+=SIZEZ-1)
    {
      A[0][i][j][k] = (double)(i*2.5 + j*3.3 + k*0.5 ) /3.0;
      test[0][i][j][k]=A[0][i][j][k];
    }
}
int check(const int iter)
{
   int i,j,k;
   int itr;
   int ii,jj,kk;
   printf("begin check\n");
       
    for(itr = 0; itr < iter; ++itr){
     for(i = 1; i < SIZEX - 1; i++)
      for( j = 1; j < SIZEY - 1; j++)
       for( k = 1; k < SIZEZ - 1; k++)
         TestDo3d7pkernel(itr,i,j,k);
      }
       
    for(ii = 1; ii < SIZEX - 1; ii++)
      for(jj = 1; jj < SIZEY - 1; jj++)
       for(kk = 1; kk < SIZEZ - 1; kk++){
          if( fabs(test[(iter+1)%2][ii][jj][kk]  - A[(iter+1)%2][ii][jj][kk]) > 1e-16 ){
             printf("i: %d   j:  %d k: %d\n",ii,jj,kk);    
             printf("test: %lf",test[(iter+1)%2][ii][jj][kk]);
             printf("but A: %lf",A[(iter+1)%2][ii][jj][kk]);
             printf("they are different!\n");               
             return -1;
         }
      }
  printf("correct\n");
  return 0;
}



int main(int argc, char** argv)
{
    int num_threads;
    int i,j,k;
    struct timeval start;
    struct timeval end;
    double total_time;       
    
    if(argc > 5)
    {
        printf("input arg error, USAGE: ./a.out nter blkx blky num_threads\n");
        return 0;
    }
    num_threads = 1;
    BLKX = 16;
    BLKY = 16;
    int ITER = 1;
    if(argc > 1)
        ITER = atoi(argv[1]);
    if(argc > 2)
        BLKX = atoi(argv[2]);
    if(argc > 3)
        BLKY = atoi(argv[3]);
    if(argc > 4)
        num_threads = atoi(argv[4]);
    Myinit();
    acemesh_runtime_init(num_threads);
    
	gettimeofday(&start,NULL);
    Do3d7p(SIZEZ, ITER, BLKX, BLKY);
    gettimeofday(&end,NULL);   
    
	acemesh_runtime_shutdown();  

    total_time = 0.0;
    total_time+=(end.tv_sec-start.tv_sec)+(end.tv_usec-start.tv_usec)/1000000.0;
    printf("total time: %lf\n", total_time);
#ifdef CHECK
    if(check(ITER))
    {
        printf("check_error\n");
        exit(1);
    }
#endif
    return 0;
}
